### ISSUE TITLE
/proc/self/uid_map: no such file or directory

### ISSUE THREAD

**Issue by gbmeuk** [joyent/smartos-live#714]
Hi,

I got redirected here from https://github.com/joyent/sdc-vmtools-lx-brand/issues/28, so I hope this is the right place.

I've been trying to run [Concourse](http://concourse.io) on Triton and hit an issue where there is no uid_map (and probably gid_map) files in /proc. I understand on Triton that this isn't ever going to be a thing. However, I do wonder whether there should be some stub code for them?

```
panic: open /proc/self/uid_map: no such file or directory

goroutine 1 [running]:
panic(0xd24700, 0xc4201d78c0)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
code.cloudfoundry.org/guardian/sysinfo.must(0x0, 0xf22f700, 0xc4201d78c0, 0xf22f700)
        /tmp/build/9674af12/garden-runc-release/src/code.cloudfoundry.org/guardian/sysinfo/max_valid_uid.go:69 +0x66
code.cloudfoundry.org/guardian/sysinfo.MustGetMaxValidUID(0xd39e00)
        /tmp/build/9674af12/garden-runc-release/src/code.cloudfoundry.org/guardian/sysinfo/max_valid_uid.go:15 +0x58
code.cloudfoundry.org/guardian/guardiancmd.init.1()
        /tmp/build/9674af12/garden-runc-release/src/code.cloudfoundry.org/guardian/guardiancmd/command.go:225 +0x37
code.cloudfoundry.org/guardian/guardiancmd.init()
        /tmp/build/9674af12/garden-runc-release/src/code.cloudfoundry.org/guardian/guardiancmd/seccomp.go:1598 +0x349
main.init()
        /tmp/build/9674af12/gopath/src/github.com/concourse/bin/cmd/concourse/worker_unix.go:24 +0xdf
```

**Comment by jjelinek** [comment]
We do not yet have support for user namespaces. This is being tracked in our internal bugtracker as OS-4483.

**Comment by gbmeuk** [comment]
Thanks @jjelinek. I can't see any comments there (probably permissions) but OS-4090 is two years old now, so is there any chance of this being implemented soon? I ask because I wonder whether generating uid_map and gid_map files is an easy small step forward that resolves issues where applications query /proc (like some Linux applications do) directly?

**Comment by jjelinek** [comment]
We have a plan to provide some level of namespace support on our roadmap, but this is a complex area and it will take time before we have anything really working. User namespaces is just one piece of the whole namespace effort. We might be able to provide some emulation for these two proc files without any other user namespace support, but it is unclear if the application would work properly or just run into another failure due to some other aspect of the namespace support which we don't yet have. I am not familiar with this application. Is there any option to run it without using namespaces?

**Comment by jclulow** [comment]
According to *user_namespaces(7)*, it would appear that there is a basic dummy file even in the initial user namespace:

> The initial user namespace has no parent namespace, but, for
> consistency, the kernel provides dummy user and group ID mapping
> files for this namespace.  Looking at the `uid_map` file (`gid_map` is
> the same) from a shell in the initial namespace shows:
>
>     $ cat /proc/$$/uid_map
>              0          0 4294967295
>
> This mapping tells us that the range starting at user ID 0 in this 
> namespace maps to a range starting at 0 in the (nonexistent) parent
> namespace, and the length of the range is the largest 32-bit unsigned
> integer.  This leaves `4294967295` (the 32-bit signed `-1` value)
> unmapped.  This is deliberate: `(uid_t)-1` is used in several
> interfaces (e.g., setreuid(2)) as a way to specify "no user ID". 
> Leaving `(uid_t)-1` unmapped and unusable guarantees that there will 
> be no confusion when using these interfaces.

**Comment by jjelinek** [comment]
Josh, thanks for the update. I opened internal bug OS-6052 for this.

**Comment by gbmeuk** [comment]
Thanks guys.  Look forward to seeing it.

**Comment by jjelinek** [comment]
Fixed in commit 4489e7e.
